### ISSUE TITLE
decrease reconnection time

### ISSUE THREAD

**Issue by balejk** [meejah/fowl#12]
I seem to observe that it takes quiet a long time for the connection to be restored after a network change (currently I have already been waiting for half an hour and still nothing). Is there anything fowl can do to decrease that delay? Or is it something that is done internally by the wormhole library?

**Comment by meejah** [comment]
Hmm, interesting.
That's basically "Dilation code, inside magic-wormhole" .. but I also maintain that library so definitely interested, although probably a tracking this via an issue on https://github.com/magic-wormhole/magic-wormhole makes more sense.

**Comment by meejah** [comment]
How long were you disconnected for?

**Comment by balejk** [comment]
> Hmm, interesting.
> That's basically "Dilation code, inside magic-wormhole" .. but I also maintain that library so definitely interested, although probably a tracking this via an issue on https://github.com/magic-wormhole/magic-wormhole makes more sense.

Yes, I was going to add that you should feel free to transfer the issue
to the other repo if appropriate, but I was thinking that the dilation
could have some reconnect timer which could be set via an argument from
the client, so it made sense to first open it here. I was not able to
find anything like it from a quick look at the dilation documentation
and even some bits of source code, though.

Could you please point me to the relevant code or at least explain how
this portion of the protocol/implementation works? I understand that the
leader of the connection should determine when the subchannel is dead,
but I was not able to find any details.

What I would naively imagine is something along the lines of SSH's
ServerAliveInterval. When the leader would deem the subchannel dead, it
would try to negotiate a new connection via the mailbox server, after
the other side has reconnected there (which I assume does happen, as
it's the central point of communication between the two clients). Or
even perhaps both sides could agree on a subchannel being dead after
they reconnect via the mailbox server.

> How long were you disconnected for?

Quite a lot, like 20 minutes maybe. I did exactly what you write in your
blog post: close the laptop, change train and reopen :-P

I have since then changed the network once more (and was offline for
even longer than before) and now I'm waiting to see if it reconnects (it
never did on the second network). This is not the first time I'm seeing
this. When I first tried fowl, I was going to raise an issue asking
about the durability of the connection as there seemed to be none and
then it just suddenly reconnected after some time. So I waited until I
test it again and since there was no improvement, I thought it better to
ask.

By the way, the other side should have remained connected the whole
time, so if I understood correctly, the mailbox server should not free
the session - right?

**Comment by meejah** [comment]
All the dilation code is here: https://github.com/magic-wormhole/magic-wormhole/tree/master/src/wormhole/_dilation

I'll take a look; it's been a bit since I've delved into the Dilation implementation.

Overall, your expectations are correct. And yes, the mailbox would typically be used to negotiate a new "generation" (I believe is the term used) of the wormhole. The Leader always decides this.

> By the way, the other side should have remained connected the whole time, so if I understood correctly, the mailbox server should not free the session - right?

Correct. BOTH sides have to be disconnected for 10 minutes before the mailbox is freed (with the default server, other deployments can choose their own timeouts).

**Comment by balejk** [comment]
> All the dilation code is here: https://github.com/magic-wormhole/magic-wormhole/tree/master/src/wormhole/_dilation

Yes, that's where I was looking briefly, but with no luck. I was hoping
you could be more specific, but I understand if you don't know either
off the top of your head.

> I'll take a look; it's been a bit since I've delved into the Dilation implementation.

I will try to take another look when I have the time, perhaps you will
beat me to it. But at least you confirm, that it has nothing to do with
fowl and is really an internal wormhole thing, if I understand
correctly.

> > By the way, the other side should have remained connected the whole time, so if I understood correctly, the mailbox server should not free the session - right?
>
> Correct. BOTH sides have to be disconnected for 10 minutes before the mailbox is freed (with the default server, other deployments can choose their own timeouts).

It's been twelve hours since my last network change and still nothing,
so I guess it's not coming back this time.

I have let both sides still running for now in case you have any
suggestions for what I should try or even if you wanted to inspect the
mailbox server to see whether it still considers the connection alive,
provided that you have access to it. But I will have to turn it off
eventually.

**Comment by balejk** [comment]
> All we'd be able to see on the mailbox server is if the mailbox itself is still allocated (but I think the interesting things are all in the _content_ of the messages).

That's what I originally meant - seeing whether the mailbox server still
thinks the connection is active. Of course the messages themselves would
be even more valuable.

To test the mailbox server, I tried allocating a code with the same
nameplate (by issuing `wormhole receive ...`) and got an error. On a
second attempt, it allocated it. So either I interrupted someone's
transfer by an accident, or the connection was indeed still there and I
probably broke it. But actually, I'm not sure whether this works like
this - is it possible to interrupt a connection by using the wrong code
after the connection has already been established? Shouldn't the
nameplate be freed after the sides verify each other using the code? (I
am not saying that they should not use the mailbox server, of course
they will still need it to arrange for the transit connection, however I
would maybe expect that the nameplate would only be relevant so that
they can find each other initially and then the mailbox server could
assign them some less human friendly identifier. But maybe that doesn't
really make sense, to use two different things for session
identification, although it could prevent these interruptions of
established connections by a MITM attacker, if those are really
possible.)

> `wormhole send --debug-state M` would be the way to dump state-transitions for the Manager machine...would probably be good to add a similar option in `fowl` for at least us to use :)

That would be really useful for debugging this.

By the way, I just unintentionally tested it again - had two devices
connected directly via a cable, both connected to the same wireless
network. I created the fowl session and then uplugged them and the
connection died and did not reconnect within maybe 20 minutes. While
those previous cases I considered rather tough (quite long and frequent
disconnects), this I would expect to reconnect.

Two questions (the first one I already kind of implicitly asked so it's
possible we won't know until one of us reads the dilation code
properly): what is the expected time to reconnect? To me those 10 to 15
minutes would seem plausible, but I'm not really sure whether there is
some reason for the leader to wait longer to see if the original
connection might come back alive (although since both sides should still
be able to communicate via the mailbox server, it would seem that there
is no reason that they can't just agree to drop the old connection (or
*generation* as you said)). In an optimal case, I would probably want a
minute or two at the maximum (but again, there may be a good reason to
wait longer before giving up on the older connection, I don't know).

Second question: were you actually able to reproduce this yourself? Or
does the durability work as expected for you?

Thank you for your patience and interest :-)

**Comment by meejah** [comment]
> But actually, I'm not sure whether this works like this - is it possible to interrupt a connection by using the wrong code after the connection has already been established?

No.

The nameplates are ephemeral; the mailboxes have 256-bit IDs and are the long-lived part. So by re-using e.g. `6-*-*` you've just re-allocated the "6" nameplate (but it will now point at a different mailbox).

**Comment by meejah** [comment]
Yes, I agree from a usability standpoint something less than 10 minutes is desirable.

I have not repeated this issue myself (see ticket on magic-wormhole, I tried an overnight experiment), but haven't spent a ton of time yet -- and my familiarity with the deep aspects of Dilation has waned a little :)

**Comment by meejah** [comment]
Note that with `fowl` there's another aspect here: since Dilation is invisible to the streams being forwarded, it's possible that client software itself has timeouts. I've been mostly experimenting with `telnet` and `nc` ... but the "higher level" protocol may play a role here too :/

**Comment by balejk** [comment]
> The nameplates are ephemeral; the mailboxes have 256-bit IDs and are the long-lived part. So by re-using e.g. `6-*-*` you've just re-allocated the "6" nameplate (but it will now point at a different mailbox).

Exactly what I would expect, thanks.

> I have not repeated this issue myself (see ticket on magic-wormhole, I tried an overnight experiment), but haven't spent a ton of time yet -- and my familiarity with the deep aspects of Dilation has waned a little :)

I find it strange that it worked for you on a local network while it did
not reconnect for me on the cable either after I plugged it back in
(although I did not wait for too long). It occurred to me that it could
have something to do with the fact that in your case it might not have
been the leader who was put to sleep, however in my case both sides
still had access to the mailbox server, so that should not be the
reason. Also, you write about only seeing `-2` messages - that denotes
the generation? In that case, could it be possible that your setup
really reused the old connection, because putting the laptop to sleep
did not close it (because the laptop had no time to handle it before
sleeping, then it woke up and the connection was still active and so it
got reused, while in my case, the network was interrupted for some time
while the device was still active, so it had time to handle that and
close the connection on its side after some time)? I'm just speculating
though.

How long did it take to reconnect for you?

> Note that with `fowl` there's another aspect here: since Dilation is invisible to the streams being forwarded, it's possible that client software itself has timeouts. I've been mostly experimenting with `telnet` and `nc` ... but the "higher level" protocol may play a role here too :/

That should not be my case - I am experimenting with SSH and I'm
starting a new session every time. That's not a problem, right, the
forward should be durable even if nothing is going through the channel
at the moment?

**Comment by meejah** [comment]
>  That's not a problem, right, the forward should be durable even if nothing is going through the channel at the moment?

Yes, that should work.

Also yeah I didn't try to figure out which side was Leader (which could indeed make a difference).

Yes, I was talking about "generations" with the `-2` thing.

**Comment by meejah** [comment]
> Yes, that should work.

...welllll, so there are network devices like NATs that like to cancel TCP connections that are "inactive". We could send pings (but I think the application has to decide to do that, not Dilation itself).

**Comment by balejk** [comment]
> ...welllll, so there are network devices like NATs that like to cancel TCP connections that are "inactive". We could send pings (but I think the application has to decide to do that, not Dilation itself).

Well OK, but I assume the mailbox connection would be considered active
and so even if the subchannels got cut off, the wormhole should be able
to reconnect them, at least on demand, no?

**Comment by meejah** [comment]
> Well OK, but I assume the mailbox connection would be considered active and so even if the subchannels got cut off, the wormhole should be able to reconnect them, at least on demand, no?

Yes.

There's some complexity in here, though: often you don't learn that some NAT "disconnected" the stream until you try to send more bytes (hence the ubiquity of sending pings). But, overall, it **should** work.

**Comment by balejk** [comment]
> The server/client distinction is arbitrary and has no intended
> correspondence to which side was the leader.

In fact, the "client" is the laptop on the "flaky" network, where I have
requested listening on local port and forwarding of the traffic to a
port on the "server" device. "server"'s connection should be quite
stable.

**Comment by meejah** [comment]
Some more annec-data: I just did an experiment, consisting of the following:

- connect laptop to desktop
- forward ssh ports
- connect ssh between desktop / laptop (over fowl / wormhole)
- above is all on a LAN / home network
- laptop sleeps
- relocate (took about 30 minutes before I regained network, on a new network)
- tried "doing a thing" in ssh (nothing)
- after about 10 minutes, i started to see _some_ action in the fowl console (it claimed to be forwarding 36 bytes at a time, occasionally -- assuming keepalive)
- after about 10-15 _additional_ minutes, ssh began working (I'm not sure _precisely_ when)

So, it did eventually re-negotiate and forward all the missing SSH bytes, but "half an hour" is of course way too long.
I don't know what affects the above (e.g. is this something a program like "fowl" can control, or is it ultimately a magic-wormhole problem?)

**Comment by meejah** [comment]
The connection survived the return trip too, but I didn't make a good note of times...

**Comment by meejah** [comment]
So, from my ad-hoc real-world testing, there is definitely at least two problems here:

- reconnection takes "too long" sometimes
- there isn't enough feedback (up to the user) about what Dilation may (or may not) be doing (e.g. "re-connecting" vs. "thinks it is connected right now" etc).

Even if the re-connection still takes as long as it does right now, having some feedback about the process would greatly help the user-experience (and presumably help debug why it takes a long time sometimes).

**Comment by balejk** [comment]
Agreed. I will try to remember to test this again with changing networks
and see if I observe the same behaviour as before (that is, no
reconnection after hours) or rather reproduce what you write (so,
reconnection within some (still-not-so-)reasonable amount of time).

**Comment by meejah** [comment]
It _might_ be related to how long you're offline too ... my next step here will be to add more instrumentation to magic-wormhole so that fowl can produce some output to the user indicating what's going on (it might be as simple as "exponential backoff, with no upper bound" or something, for example)

**Comment by balejk** [comment]
That would sound plausible, but see what I wrote above about my flaky
network. Plus, if I recall correctly, you put the laptop to sleep over
night and it reconnected within 30 minutes.

Either way, adding some verbose connection status reporting to
magic-wormhole seems like the best next step.

**Comment by balejk** [comment]
I tested the "get on a train" scenario again and something somewhat
interesting happened. I started two fowl sessions on the same device
(within about fifteen minutes, simply because I forgot to specify some
-L/-R connections on the first one) and connected from a second device.
The second device was subsequently offline for about an hour and then
connected to a different network. One of the two fowl sessions
reconnected in 20 minutes, the other still has not (it's been many
hours).

Do you have any explanation for this? It doesn't sound much like the
exponential backoff, does it? The only thing that comes to mind which
could be different between the two sessions is which device was the
leader for each. Or is there something else too?

**Comment by meejah** [comment]
I wouldn't expect leader/follower from purely the `fowl` level to have any effect (there is a Leader / Follower at the Dilation level, but it's random depending on the keys).

**Comment by balejk** [comment]
> Is the session that _didn't_ connect also the session with no subchannels? With no traffic, the OS can often pretend that a TCP stream is alive (until it tries to send a byte, and then discovers that it is not).

I was not using either connection at the time when I lost the first
network and I only attempted to use it once I connected to the other
network. So no subchannels should have been open for either connection.

> I wouldn't expect leader/follower from purely the `fowl` level to have any effect (there is a Leader / Follower at the Dilation level, but it's random depending on the keys). 

I meant Leader/Follower at the Dilation level.
